packages/babel-core/src/transformation/internal-plugins/shadow-functions.js

```diff
@@ -38,21 +40,37 @@ function remap(path, key, create) {
     }
 
     if (path.isProgram()) {
+      passedShadowFunction = true;
+
       return true;
-    } else if (path.isFunction()) {
+    } else if (path.isFunction() && !path.isArrowFunctionExpression()) {
       if (shadowFunction) {
-        return path === shadowFunction || path.node === shadowFunction.node;
+        if (path === shadowFunction || path.node === shadowFunction.node) return true;
       } else {
-        return !path.is("shadow");
+        if (!path.is("shadow")) return true;
       }
+
+      passedShadowFunction = true;
+      return false;
     }
 
     return false;
   });
 
+  if (shadowFunction && fnPath.type === "Program" && shadowFunction.type !== "Program"){
```

Comment:
We should use the `t.is` helpers here, at least for consistency. We should seldom touch `.type`.